### DISCUSSION TITLE
file dialog with path having cjk character

### DISCUSSION THREAD

**Discussion by cstkingkey** [fltk-rs/fltk-rs#1078] [Q&A]
can file dialog handle cjk path?
getting invalid characters

**Comment by MoAlyousef** [comment]
FileDialog is the system’s file dialog, so the behavior should be the same. So does it concern paths you’re creating via the dialog, paths you’re passing programmatically in your Rust code, paths which already exist but can’t access?

**Reply by cstkingkey** [reply]
to_string_lossy
`   ///Converts a `CStr` into a <code>[Cow]<[str]></code>.
    ///
    /// If the contents of the `CStr` are valid UTF-8 data, this
    /// function will return a <code>[Cow]::[Borrowed]\(&[str])</code>
    /// with the corresponding <code>&[str]</code> slice. Otherwise, it will
    /// replace any invalid UTF-8 sequences with
    /// [`U+FFFD REPLACEMENT CHARACTER`][U+FFFD] and return a
    /// <code>[Cow]::[Owned]\(&[str])</code> with the result.`

But CStr is for single and double-byte character sets. DBCS may need 2 bytes to represent a single character, whereas UNICODE always uses two bytes per character. 
If I'm reading rightly, some characters can not be converted correctly.

**Reply by MoAlyousef** [reply]
That doesn’t answer any of my questions

**Reply by cstkingkey** [reply]
when selecting file, the result filename is invalid. the Chinese character is not shown correctly. can not be used to open the file.
selecting a file with pure ascii character is working fine

**Reply by MoAlyousef** [reply]
Do all filenames with Chinese characters fail?
My filesystems has Arabic files and dirs and things work correctly.

Trying on my system (windows and linux) on a file with Chinese characters:
``` rust
use fltk::{app, dialog, button::Button, frame::Frame, prelude::*, window::Window};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    let mut frame = Frame::default().with_size(200, 100).center_of(&wind);
    let mut but = Button::new(160, 210, 80, 40, "Click me!");
    wind.end();
    wind.show();

    but.set_callback(move |_| {
        let mut dlg = dialog::FileDialog::new(dialog::FileDialogType::BrowseSaveFile);
        dlg.show();
        frame.set_label(&dlg.filename().to_str().unwrap());
    });

    app.run().unwrap();
}
```

![image](https://user-images.githubusercontent.com/37966791/147586808-3bf8b9e0-95bd-43ac-b4b8-01e08387191f.png)

Works also for getting the content of the file.

For unpaired surrogates, it seems even Rust can't deal correctly with it
https://stackoverflow.com/questions/46342644/how-can-i-get-a-path-from-a-raw-c-string-cstr-or-const-u8

If you have a test case that I can try, I can try to debug your issue.

**Reply by MoAlyousef** [reply]
If you have a suggestion to improve the FileDialog::filename() method to return unpaired surrogates, you can suggest them here or open a PR.

**Comment by cstkingkey** [comment]
thanks for your time.
